demo_nodes_cpp/src/parameters/set_parameters_callback.cpp

```diff
@@ -0,0 +1,118 @@
+// Copyright 2022 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * Example usage: changing param1 successfully will result in setting of param2.
+ * ros2 service call /set_parameters_callback/set_parameters rcl_interfaces/srv/SetParameters
+                    "{parameters: [{name: "param1", value: {type: 3, double_value: 1.0}}]}"
+ */
+
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "rcl_interfaces/msg/set_parameters_result.hpp"
+#include "rclcpp_components/register_node_macro.hpp"
+
+#include "demo_nodes_cpp/visibility_control.h"
+
+/**
+ * node for demonstrating correct usage of pre_set, on_set
+ * and post_set parameter callbacks
+ */
+namespace demo_nodes_cpp
+{
+class SetParametersCallback : public rclcpp::Node
+{
+public:
+  DEMO_NODES_CPP_PUBLIC
+  explicit SetParametersCallback(const rclcpp::NodeOptions & options)
+  : Node("set_parameters_callback", options)
+  {
+    this->declare_parameter<double>("param1", 1.0);
+    this->declare_parameter<double>("param2", 2.0);
+
+    // tracks "param1" value
+    internal_tracked_class_parameter_1_ = this->get_parameter("param1").as_double();
+
+    // tracks "param2" value
+    internal_tracked_class_parameter_2_ = this->get_parameter("param2").as_double();
```

Comment:
I think this whole thing can be:

```suggestion
    // tracks "param1" value
    internal_tracked_class_parameter_1_ = this->declare_parameter("param1", 1.0);
    // tracks "param2" value
    internal_tracked_class_parameter_2_ = this->declare_parameter("param2", 2.0);
```